modules/pam_setquota/pam_setquota.c

```diff
@@ -0,0 +1,317 @@
+             p->dqb_ihardlimit);
+}
+
+static int _pam_parse_dqblk(int argc, const char **argv, struct if_dqblk *p) {
+  int bhard = FALSE, bsoft = FALSE, ihard = FALSE, isoft = FALSE;
+
+  /* step through arguments */
+  for (; argc-- > 0; ++argv) {
+    const char *str;
+    char *ep = NULL;
+    if ((str = pam_str_skip_prefix(*argv, "bhardlimit=")) != NULL) {
+      p->dqb_bhardlimit = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_bhardlimit < 0))
+        p->dqb_bhardlimit = 0;
+      p->dqb_valid |= QIF_BLIMITS;
+      bhard = TRUE;
+    } else if ((str = pam_str_skip_prefix(*argv, "bsoftlimit=")) != NULL) {
+      p->dqb_bsoftlimit = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_bsoftlimit < 0))
+        p->dqb_bsoftlimit = 0;
+      p->dqb_valid |= QIF_BLIMITS;
+      bsoft = TRUE;
+    } else if ((str = pam_str_skip_prefix(*argv, "ihardlimit=")) != NULL) {
+      p->dqb_ihardlimit = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_ihardlimit < 0))
+        p->dqb_ihardlimit = 0;
+      p->dqb_valid |= QIF_ILIMITS;
+      ihard = TRUE;
+    } else if ((str = pam_str_skip_prefix(*argv, "isoftlimit=")) != NULL) {
+      p->dqb_isoftlimit = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_isoftlimit < 0))
+        p->dqb_isoftlimit = 0;
+      p->dqb_valid |= QIF_ILIMITS;
+      isoft = TRUE;
+    } else if ((str = pam_str_skip_prefix(*argv, "btime=")) != NULL) {
+      p->dqb_btime = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_btime < 0))
+        p->dqb_btime = 0;
+      p->dqb_valid |= QIF_BTIME;
+    } else if ((str = pam_str_skip_prefix(*argv, "itime=")) != NULL) {
+      p->dqb_itime = strtol(str, &ep, 10);
+      if (!ep || (p->dqb_itime < 0))
+        p->dqb_itime = 0;
+      p->dqb_valid |= QIF_ITIME;
+    }
+  }
+
+  /* return FALSE if a softlimit or hardlimit has been set
+   * independently of its counterpart.
+   */
+  return !(bhard ^ bsoft) && !(ihard ^ isoft);
+}
+
+static void _pam_parse_params(int argc, const char **argv,
+                              struct pam_params *p) {
+  /* step through arguments */
+  for (; argc-- > 0; ++argv) {
+    const char *str;
+    char *ep = NULL;
+    if ((str = pam_str_skip_prefix(*argv, "startuid=")) != NULL) {
+      p->start_uid = strtol(str, &ep, 10);
+      if (!ep || (p->start_uid < 0) || (p->start_uid >= MAX_UID_VALUE))
+        p->start_uid = PAM_USERTYPE_UIDMIN;
+    } else if ((str = pam_str_skip_prefix(*argv, "enduid=")) != NULL) {
+      p->end_uid = strtol(str, &ep, 10);
+      if (!ep || (p->end_uid < 0) || (p->end_uid >= MAX_UID_VALUE))
+        p->end_uid = 0;
+    } else if ((str = pam_str_skip_prefix(*argv, "fs=")) != NULL) {
+      /* Remove the unnecessary '/' from the end of fs parameter */
+      char *temp = NULL;
+      size_t len = strlen(str);
+      temp = strdup(str);
+      if (len > 1) {
+        if (temp[len - 1] == '/') {
+          temp[len - 1] = '\0';
+        }
+      }
+      p->fs = strdup(temp);
+      free(temp);
```

Comment:
I don't see why you need so many copies of the string.
I'm not sure whether you really need to make a copy, but if you do, please do not forget to check return code and free the copy afterwards.

Reply:
This is me fumbling with strings in C, assistance is appreciated.
str is const char here, so I can't remove the / from the end, right? So I need a temp variable to do the changes and then strdup it into p->fs, because just p->fs = temp will cause p->fs to contain garbage after free(temp).

Reply:
The only place where you use param.fs as a string is
```    } else if ((strcmp(param.fs, mnt->mnt_dir) == 0) ||
               (strcmp(param.fs, mnt->mnt_fsname) == 0)) {
```

If you could rewrite it without strcmp, then you could save e.g. ```param.fs``` and ```param.fs_len``` and use them instead.

Reply:
Sorry, either it is too late or I have reached the limit of my understanding of C.

Reply:
Hmm, something like this?
```
struct pam_params {
   uid_t start_uid;
   uid_t end_uid;
   const char *fs;
  size_t fs_len;
   int overwrite;
   int debug;
 };
```
and then

```
} else if ((str = pam_str_skip_prefix(*argv, "fs=")) != NULL) {
      p->fs = str;
      p->fs_len = strlen(str);
      /* Mask the unnecessary '/' from the end of fs parameter */
      if (p->fs_len > 1 && p->fs[len - 1] == '/')
        --p->fs_len;
}
```
and finally

```
    } else if ((strncmp(param.fs, mnt->mnt_dir, param.fs_len) == 0) ||
               (strncmp(param.fs, mnt->mnt_fsname, param.fs_len) == 0)) {
```
This is all I can come up right now at this time.

Reply:
> ```
>     } else if ((strncmp(param.fs, mnt->mnt_dir, param.fs_len) == 0) ||
>                (strncmp(param.fs, mnt->mnt_fsname, param.fs_len) == 0)) {
> ```

```
} else if ((strncmp(param.fs, mnt->mnt_dir, param.fs_len) == 0
                    && mnt->mnt_dir[param.fs_len] == '\0') ||
           (strncmp(param.fs, mnt->mnt_fsname, param.fs_len) == 0
                    && mnt->mnt_fsname[param.fs_len] == '\0')) {
```

Reply:
Interesting, yes.
Checking for the string terminator to check if the match would be exact on the 2nd paramater to strncmp was something not obvious to me.